azurelinuxagent/common/protocol/wire.py

```diff
@@ -647,7 +647,7 @@ def fetch(self, uri, headers=None, use_proxy=None, decode=True):
         logger.verbose("Fetch [{0}] with headers [{1}]", uri, headers)
         content = None
         response = self._fetch_response(uri, headers, use_proxy)
-        if response is not None:
+        if not restutil.request_failed(response):
```

Comment:
even with this change we still don't handle the error in some paths. now instead of returning a reponse with an error we return None. If you follow the call stack upwards you will see paths that ignore the None value and do not report the error.

Reply:
So, the error on the request level will always be logged as a warning since this is checked in `azurelinuxagent.common.protocol.wire.WireClient._fetch_response` (called just above this line).

Inspecting the scenarios that rely on `azurelinuxagent.common.protocol.wire.WireClient.fetch`:
1) Fetching manifest: an ExtensionDownloadError will be raised since the response will not be None instead of failed response body. The exception is raised in `azurelinuxagent.common.protocol.wire.WireClient.fetch_manifest` if response is None.
2) Download extension handler package: an exception will be raised due to:
```
if not self.protocol.download_ext_handler_pkg(source_uri, target_file):
    raise Exception("Failed to download extension package - no error information is available")
```
3) Get artifacts profile: there is a check against None values in `azurelinuxagent.ga.exthandlers.ExtHandlersHandler._extension_processing_allowed`:
```
artifacts_profile = self.protocol.get_artifacts_profile()
if artifacts_profile and artifacts_profile.is_on_hold():
```

So the handling of None is already covered. However, I agree the error reporting can be improved, since we only have the warning on the request level and then all the way on the top of the call stack in cases 1) and 2), and nothing in 3). I'm adding an error log in the middle functions, `azurelinuxagent.common.protocol.wire.WireClient.get_artifacts_profile`, `azurelinuxagent.common.protocol.wire.WireClient.fetch_manifest` and `azurelinuxagent.common.protocol.wire.WireProtocol.download_ext_handler_pkg`.

Reply:
One thing that worries me about adding a log warning when requests fail in `azurelinuxagent.common.protocol.wire.WireClient.get_artifacts_profile` is that we call this method every 3 seconds. However, even today we have this potential verbosity in case of (consistent) failure since the failed requests would be logged lower in the call stack, in `azurelinuxagent.common.protocol.wire.WireClient._fetch_response`.

Reply:
let's sync offline on this

Reply:
Suggestion: Since you're touching this code anyways, why not make the 3 calls consistent? Like make sure to raise an error if download fails and then handle them in their callers to reduce ambiguity. 
This is definitely a bigger change than what you started with but since you're touching the code anyways might as well make some efforts into solving it permanently :)